### PR TITLE
Fixes #21160: Upgrade to React@16

### PR DESCRIPTION
**Pull request by danseethaler**
The current Foreman code is compliant with React@15.6.2
which means everything should work with React@16.

I've looked through all the npm peer dependency errors
and made a status report for each:

- react-komposer - PR open [here](https://github.com/arunoda/react-komposer/pull/169)
- radium - PR open [here](https://github.com/FormidableLabs/radium/pull/930)
- react-addons-test-utils - moved to `react-dom/test-utils`
- react-icons - PR open [here](https://github.com/gorangajic/react-icons/pull/117)
- react-modal - No pending PR - likely a good idea to use a package that utilizes the React@16 portal functionality. [this may work](https://www.npmjs.com/package/react-portal)
- react-simple-di - PR open [here](https://github.com/kadirahq/react-simple-di/pull/13)
- react-test-renderer - This is part of React itself
- react-transition-group - Upgraded to 1.2.1 to support React@16
- react-treebeard - 2.0.3 should accept React@16. not sure why we're getting a peer dependency error
- react-stubber - PR open [here](https://github.com/kadirahq/react-stubber/pull/4)

http://projects.theforeman.org/issues/21160

### CODE REVIEW

**Comment by ohadlevy** [comment]
Thanks @danseethaler as expected, please now the test failures due to enzyme changes

**Comment by waldenraines** [comment]
As discussed in https://github.com/theforeman/foreman/pull/4733#issuecomment-333320581 I think waiting a week or so for other libraries to update to v16 before merging this is a good call but this will at least get us unblocked to try out portals.

**Comment by danseethaler** [comment]
I've jumped into fixing the enzyme testing errors and found that enzyme also needs to be updated to 3.0.0 which brings some breaking changes. I'm working through the `.test.js` files to fix the breaks and will submit my changes when ready.

**Comment by danseethaler** [comment]
After taking a look at the enzyme upgrade and the current tests it seems like it would make sense to use jest's snapshot feature in some places where we're evaluating component structure. Enzyme changed their `.children()` api and it has broken a number of tests. The tests that it did break seems like they would be much better suited to snapshot test anyway. I'm not sure if our team has used the jest snapshot feature before but it has a ton of advantages like clean git diffs, and much easier test creation and maintenance. [See more here](https://facebook.github.io/jest/docs/en/snapshot-testing.html)

I'm happy to work on integrating snapshots into the test suite to get everything to a passing state.  I've already started trying it out and it works great. It may make sense to upgrade enzyme in a separate PR as a precursor to React@16.

@waldenraines @ohadlevy let me know your thoughts.

**Comment by ohadlevy** [comment]
:+1: from my side

**Comment by waldenraines** [comment]
> @waldenraines @ohadlevy let me know your thoughts.

I'm for it, I've used snapshot in another project and it was useful.

**Comment by danseethaler** [comment]
Currently blocked by #4887. Will rebase off master and update after that PR is merged.

**Comment by ohadlevy** [comment]
https://blog.discordapp.com/lessons-from-migrating-a-large-codebase-to-react-16-e60e49102aa6 had some nice tips

**Comment by ohadlevy** [comment]
And.. please rebase :)

**Comment by danseethaler** [comment]
This is good to go from my perspective. However, I'm not sure what parts of foreman are currently using React. @ohadlevy do you mind running this and looking at react parts of foreman to see if we have any errors or issues?

**Comment by ohadlevy** [comment]
@danseethaler when installing, i see the following:
```
├─┬ enzyme-adapter-react-16@1.0.1 
│ └── lodash@4.17.4 
├─┬ UNMET PEER DEPENDENCY react@16.0.0
│ └── fbjs@0.8.16 
├─┬ UNMET PEER DEPENDENCY react-dom@16.0.0
│ └── fbjs@0.8.16 
└─┬ UNMET PEER DEPENDENCY react-test-renderer@16.0.0
  └── fbjs@0.8.16 
```

probably need to update version of lodash and fbjs?

**Comment by ohadlevy** [comment]
also, all of the following warnings are shown:
```

npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.0.0 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN @storybook/react-fuzzy@0.4.0 requires a peer of react@^0.14.7 || ^15.0.0 but none was installed.
npm WARN enzyme-adapter-react-15@1.0.1 requires a peer of react@^15.5.0 but none was installed.
npm WARN enzyme-adapter-react-15@1.0.1 requires a peer of react-dom@^15.5.0 but none was installed.
npm WARN enzyme-adapter-react-15@1.0.1 requires a peer of react-test-renderer@^15.5.0 but none was installed.
npm WARN react-komposer@1.13.1 requires a peer of react@^0.14.3 || ^15.0.0 but none was installed.
npm WARN patternfly-react@0.6.2 requires a peer of react@^15.5.0 but none was installed.
npm WARN radium@0.19.4 requires a peer of react@^15.3.0 but none was installed.
npm WARN react-addons-test-utils@15.6.2 requires a peer of react-dom@^15.4.2 but none was installed.
npm WARN react-icons@2.2.5 requires a peer of react@^0.14.0 || ^15.0.0 but none was installed.
npm WARN react-icons@2.2.5 requires a peer of react-dom@^0.14.0 || ^15.0.0 but none was installed.
npm WARN react-komposer@2.0.0 requires a peer of react@^0.14.7 || ^15.0.0 but none was installed.
npm WARN react-modal@2.3.2 requires a peer of react@^0.14.0 || ^15.0.0 but none was installed.
npm WARN react-modal@2.3.2 requires a peer of react-dom@^0.14.0 || ^15.0.0 but none was installed.
npm WARN react-simple-di@1.2.0 requires a peer of react@^0.14.0 || ^15.0.0 but none was installed.
npm WARN react-stubber@1.0.0 requires a peer of react@^0.14.7 || ^15.0.0 but none was installed.
npm WARN react-treebeard@2.0.3 requires a peer of react@^15.5.4 but none was installed.
npm WARN react-treebeard@2.0.3 requires a peer of react-dom@^15.5.4 but none was installed.
```

**Comment by danseethaler** [comment]
@ohadlevy just wanted to ping you on this one. I'm not sure exactly how long we want to wait for the community package updates but the unmet dependencies are shrinking and I don't see any outstanding issues. Let me know what you think.

**Comment by ohadlevy** [comment]
@danseethaler how did you get this pr going? every time i try it i get npm errors?

also, shouldn't we update storybook to a compatible version?

**Comment by danseethaler** [comment]
@ohadlevy I do get some warnings when doing `npm install` but no errors. Have you done a `rm -rf node_modules`? I'll look into the react-storybook as well; we' should probably update.

**Comment by danseethaler** [comment]
@ohadlevy It looks like the storybook dependency that's complaining is storybook/react-fuzzy. There's an [open pr](https://github.com/storybooks/react-fuzzy-search/pull/3) for the React@16 peerDependency bump. Let me know if you have any trouble doing the `npm install` after removing the `node_modules` folder.

**Comment by ohadlevy** [comment]
@danseethaler removing node modules did solve it, I wonder why it can't simply upgrade (and probably going to be annoying for every developer). is it because of the wrong dependencies ?

**Comment by waldenraines** [comment]
> @danseethaler removing node modules did solve it, I wonder why it can't simply upgrade (and probably going to be annoying for every developer). is it because of the wrong dependencies ?

It would have been interesting to do an `npm prune` prior to see if that fixed things.  Removing `node_modules` is something I have had to do several times in this project and others throughout my career.  Node is far from perfect and I think those with ample experience with node know just to try removing `node_modules` and to try again.

At a frustration level, to me, it's not any more frustrating than:  
```
bundle install
rm Gemfile.lock
bundle install
foreman start
<Ctrl+c>
rake db:migrate
foreman start
*finally do things*
```

**Comment by ohadlevy** [comment]
> It would have been interesting to do an npm prune prior to see if that fixed things. Removing node_modules is something I have had to do several times in this project and others throughout my career. Node is far from perfect and I think those with ample experience with node know just to try removing node_modules and to try again.
> 
I would suggest trying out yarn, it usually does handle issues such as this one, its sadly that one of our depdenencies broke it for so we had to temporarly stop using yarn (https://github.com/DataTables/Dist-DataTables-Bootstrap/pull/2)

> At a frustration level, to me, it's not any more frustrating than:
> 
> bundle install
> rm Gemfile.lock

TBH: I can't remember when was the last time i had to do that.

**Comment by waldenraines** [comment]
> TBH: I can't remember when was the last time i had to do that.

This week for me 😄   but again it's not a big deal to me.  I feel like there is always some "hardhat area" involved when you are developing code as developer tools don't always work perfectly. 

But yeah I would definitely be in favor of trying to use yarn again.  I have seen it work well in other projects.

**Comment by danseethaler** [comment]
> I would suggest trying out yarn

I'd definitely be up for yarn, but I also wonder about just upgrading to npm@5 and using a `package-lock.js`. My box is running npm@3 and I know a lot of issues has been fixed since then. I assume there's some configuration in ansible that's installing npm@3. @ohadlevy @waldenraines  Any thoughts on updating the default npm to be version 5?

**Comment by waldenraines** [comment]
> I'd definitely be up for yarn, but I also wonder about just upgrading to npm@5 and using a package-lock.js. My box is running npm@3 and I know a lot of issues has been fixed since then. I assume there's some configuration in ansible that's installing npm@3. @ohadlevy @waldenraines Any thoughts on updating the default npm to be version 5?

I'm not opposed to it.  Probably after 1.17 is released would be a good time to do it so we have a full release cycle to ensure we didn't break anything.  I added it to our backlog.

**Comment by waldenraines** [comment]
@danseethaler looks like this needs a rebase.

**Comment by danseethaler** [comment]
> @danseethaler looks like this needs a rebase.

all set 👍

**Comment by waldenraines** [comment]
@ohadlevy what else would you like to see before this is merged?

**Comment by ohadlevy** [comment]
@waldenraines I think(?) that if we merge it when dependencies are not fully working, it would break nightly again? 

Also, the tests are failing.

**Comment by danseethaler** [comment]
> Also, the tests are failing.

I'll take a look at the tests.

File: package.json [hunk 1]
```diff
@@ -65,9 +65,10 @@
     "lodash": "~4.15.0",
     "multiselect": "~0.9.12",
     "prop-types": "^15.6.0",
-    "react": "^15.6.2",
+    "raf": "^3.4.0",
```
Comment (ohadlevy) [inline package.json:68]: just wondering - why is this required?

Comment (danseethaler) [reply package.json:68]: It's a polyfill for `requestAnimationFrame` that React 16 relies on. `raf` is recommended since it supports the polyfill in jest testing that is also required.

Here's a link to the v16 release notes that talks about it https://reactjs.org/blog/2017/09/26/react-v16.0.html#javascript-environment-requirements

**Comment by danseethaler** [comment]
> I think(?) that if we merge it when dependencies are not fully working, it would break nightly again?

@ohadlevy @waldenraines I don't know how the nightly build process works but I would point out that we already have deprecated package warnings in foreman with `npm install`. I would assume that peer-dependency warnings wouldn't break anything if the install completes successfully - which it does.

**Comment by ohadlevy** [comment]
@danseethaler let's see :) thanks!
